src/main/java/com/aws/greengrass/clientdevices/auth/iot/registry/ThingRegistry.java

```diff
@@ -62,16 +68,46 @@ public boolean isThingAttachedToCertificate(Thing thing, Certificate certificate
     }
 
     private void registerCertificateForThing(Thing thing, Certificate certificate) {
-        registry.put(thing.getThingName(), certificate.getIotCertificateId());
+        Set<CertificateEntry> certEntrySet = registry.computeIfAbsent(thing.getThingName(), s -> new HashSet<>());
+        // replace previous association with a new entry with extended TTL
+        clearThingCertificateAssociation(thing, certificate);
+        certEntrySet.add(getNewCertificateEntry(certificate));
+        registry.put(thing.getThingName(), certEntrySet);
     }
 
-    private void clearRegistryForThing(Thing thing) {
-        registry.remove(thing.getThingName());
+    private void clearThingCertificateAssociation(Thing thing, Certificate certificate) {
+        Set<CertificateEntry> certSet = registry.get(thing.getThingName());
+        if (certSet != null) {
+            certSet.removeIf(certificateEntry ->
+                    certificateEntry.getIotCertificateId().equals(certificate.getIotCertificateId()));
+        }
     }
 
     private boolean isCertificateRegisteredForThing(Thing thing, Certificate certificate) {
-        return Optional.ofNullable(registry.get(thing.getThingName()))
-                .filter(certId -> certId.equals(certificate.getIotCertificateId()))
-                .isPresent();
+        Set<CertificateEntry> certSet = registry.get(thing.getThingName());
+        if (certSet != null) {
+            return certSet.stream()
+                    .filter(this::isValidCertificateEntry)
+                    .anyMatch(certEntry -> certEntry.getIotCertificateId().equals(certificate.getIotCertificateId()));
+        }
+        return false;
+    }
+
+    private CertificateEntry getNewCertificateEntry(Certificate certificate) {
+        return new CertificateEntry(Instant.now().plusSeconds(RegistryConfig.REGISTRY_CACHE_ENTRY_TTL_SECONDS),
+                null, certificate.getIotCertificateId());
+    }
+
+    private boolean isValidCertificateEntry(CertificateEntry certEntry) {
+        return certEntry.getValidTill().isAfter(Instant.now());
+    }
+
+    /**
+     * Removes stale (invalid) registry entries.
+     */
+    void refreshRegistry() {
+        registry.values().forEach(certificateEntries ->
+                certificateEntries.removeIf(certEntry -> !isValidCertificateEntry(certEntry)));
+        registry.values().removeIf(Set::isEmpty);
```

Comment:
doesn't look like this does anything? did you mean to remove the entry instead?

Reply:
The removal on #L109 might leave an entry with an empty-set value in the registry map: `(String, Set<CertificateEntry>)`. So we remove such unnecessary empty-set entries.

Reply:
Ah, i didn't realize removal here takes care of removing the full entry. nevermind :)